src/vs/editor/contrib/stickyScroll/browser/stickyScrollModelProvider.ts

```diff
@@ -0,0 +1,433 @@
+
+		this._store.clear();
+		this._updateScheduler = new Delayer<StickyModel | null>(300); // TODO: What values to use?
+		this._store.add({
+			dispose: () => {
+				this._cancelModelPromise();
+				this._updateScheduler?.cancel();
+				this._updateScheduler = null;
+			}
+		});
+		this._cancelModelPromise();
+
+		return await this._updateScheduler.trigger(async () => {
+
+			for (const modelProvider of this._modelProviders) {
+
+				const _stopWatch = new StopWatch(true);
+				if (modelProvider.provider) {
+					this._updateDebounceInfo = this._languageFeatureDebounceService.for(modelProvider.provider, 'Sticky Scroll', { min: 200 });
+				}
+
+				const { statusPromise, modelPromise } = modelProvider.computeStickyModel(
+					textModel,
+					textModelVersionId,
+					token
+				);
+				this._modelPromise = modelPromise;
+				const status = await statusPromise;
+				switch (status) {
+					case Status.CANCELED:
+						this._store.clear();
+						return null;
+					case Status.VALID:
+						if (this._updateScheduler && this._updateDebounceInfo) {
+							this._updateScheduler.defaultDelay = this._updateDebounceInfo.update(textModel, _stopWatch.elapsed());
+						}
+						return modelProvider.stickyModel;
+				}
+			}
+			return null;
+		});
+	}
+}
+
+interface IStickyModelCandidateProvider<T> {
+	get stickyModel(): StickyModel | null;
+
+	get provider(): LanguageFeatureRegistry<object> | null;
+
+	/**
+	 * Method which computes the sticky model and returns a status to signal whether the sticky model has been successfully found
+	 * @param textmodel text-model of the editor
+	 * @param modelVersionId version ID of the text-model
+	 * @param token cancellation token
+	 * @returns a promise of a status indicating whether the sticky model has been successfully found as well as the model promise
+	 */
+	computeStickyModel(textmodel: ITextModel, modelVersionId: number, token: CancellationToken): { statusPromise: Promise<Status> | Status; modelPromise: CancelablePromise<T | null> | null };
+}
+
+abstract class StickyModelCandidateProvider<T> implements IStickyModelCandidateProvider<T> {
+
+	private _providerModelPromise: CancelablePromise<T> | null = null;
+	protected _stickyModel: StickyModel | null = null;
+
+	constructor() { }
+
+	get stickyModel(): StickyModel | null {
+		return this._stickyModel;
+	}
+
+	private _invalid(): Status {
+		this._stickyModel = null;
+		return Status.INVALID;
+	}
+
+	public abstract get provider(): LanguageFeatureRegistry<object> | null;
+
+	public computeStickyModel(textModel: ITextModel, modelVersionId: number, token: CancellationToken): { statusPromise: Promise<Status> | Status; modelPromise: CancelablePromise<T | null> | null } {
+		if (!this.isProviderValid(textModel)) {
+			return { statusPromise: this._invalid(), modelPromise: null };
+		}
+		this._providerModelPromise = createCancelablePromise(token => this.createModelFromProvider(textModel, modelVersionId, token));
```

Comment:
I think `_providerModelPromise` doesn't need to be a field. It seems only be used in `computeStickyModel`